{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "0cf6eb54",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pandas as pd\n",
    "import pandas as pd\n",
    "from complexcgr import CGR\n",
    "from pathlib import Path\n",
    "# Instantiate class CGR\n",
    "cgr = CGR()\n",
    "\n",
    "def get_last_name(file_path):\n",
    "    return Path(file_path).stem\n",
    "\n",
    "def preprocessing(snp_list,df_bio,df_sra,country):\n",
    "    # Step1: merge SNPs of all samples\n",
    "    with open(snp_list, \"r\") as f:\n",
    "        sample_files = f.read().splitlines()\n",
    "\n",
    "    try:\n",
    "        # Read the data from the first sample\n",
    "        first_data = pd.read_csv(sample_files[0], sep=\" \", header=None)\n",
    "        first_data.columns = [\"genomic_position\", \"REF\", get_last_name(sample_files[0])]\n",
    "        # Drop the REF column\n",
    "        first_data = first_data.drop(columns=[\"REF\"])\n",
    "\n",
    "        # Start with first_data as the merged data\n",
    "        merge_data = first_data\n",
    "\n",
    "        # add other samples and merge them one by one\n",
    "        for sample in sample_files[1:]:\n",
    "            try:\n",
    "                pd.read_csv(sample, sep=\" \", header=None)\n",
    "#                 print(sample)\n",
    "                sample_data = pd.read_csv(sample, sep=\" \", header=None)\n",
    "                column_name = get_last_name(sample)\n",
    "                sample_data.columns = [\"genomic_position\", \"REF\", column_name]\n",
    "                # Drop the REF column\n",
    "                sample_data = sample_data.drop(columns=[\"REF\"])\n",
    "                # merge data on genomic positions\n",
    "                merge_data = merge_data.merge(sample_data, on=\"genomic_position\", how=\"outer\")\n",
    "            except:\n",
    "                continue\n",
    "\n",
    "        merge_data.set_index(\"genomic_position\", inplace=True)\n",
    "        # Tranpose the df (reverse) to have samples as rows and genomic positions as columns\n",
    "        merge_data = merge_data.T\n",
    "        #replace nan with merged data\n",
    "#        df=df.merge_data(\"N\")\n",
    "        df=merge_data.fillna(\"N\")\n",
    "        \n",
    "        df=df.reset_index()\n",
    "        df=df.rename(columns={\"index\":\"sra_accession\"})\n",
    "        \n",
    "        # Step 2: Remove constant columns\n",
    "        non_constant_cols = df.columns[df.nunique() > 1]\n",
    "        df_constant_dropped = df[non_constant_cols]\n",
    "        df_constant_dropped\n",
    "        df_constant_dropped.insert(1, \"country\", country)\n",
    "        # read metadata\n",
    "        df_bio=pd.read_table(df_bio)\n",
    "        df_sra=pd.read_table(df_sra,header=None)\n",
    "        #map sra to biosample\n",
    "        df_sra_biosample=pd.concat([df_bio,df_sra], axis=1)\n",
    "         # concat the sra and df\n",
    "        df_sra_biosample.dropna(inplace=True)\n",
    "        df_country=pd.concat([df_sra_biosample,df],axis=1)\n",
    "        df_country\n",
    "        \n",
    "        df_constant_dropped.to_csv(f\"../data/{country}_processed.csv\")\n",
    "\n",
    "        # Move the return statement outside the loop, after all samples have been processed and merged.\n",
    "        return df_constant_dropped\n",
    "\n",
    "    except Exception as e:\n",
    "        print(\"Error:\", str(e))\n",
    "        return None\n",
    "\n",
    "\n",
    "# Call the preprocessing function and store the result in the 'df' variable\n",
    "#     merge_data = merge_data.fillna(\"N\")\n",
    "#     merge_data = merge_data.replace({\"A\": 1, \"G\": 2, \"C\": 3, \"T\": 4, \"N\": 0})\n",
    "#     merge_data = merge_data.rename(columns={\"V1\": \"POS\"})\n",
    "#     tinput = merge_data.T\n",
    "#     tinput.columns = tinput.iloc[0]\n",
    "#     tinput = tinput.iloc[1:]\n",
    "#     value = len(sample_files) // 5 * 4\n",
    "#     tinput4 = tinput[tinput[\"Num\"] < value]\n",
    "#     tinput5 = tinput4.T\n",
    "#     tinput5.to_csv(\"snp_input.csv\", index=False)\n",
    "#     # FCGR Encoding using kaos package\n",
    "#     data = pd.read_csv(\"snp_input.csv\")\n",
    "#     data = data.replace({0: \"N\", 1: \"A\", 2: \"G\", 3: \"C\", 4: \"T\"})\n",
    "\n",
    "#     for i in range(900):\n",
    "#         cgr_list = cgr(data[str(i)], seq_base=[\"N\", \"A\", \"G\", \"C\", \"T\"], res=200)\n",
    "#         cgr_matrix = cgr_list[\"matrix\"]\n",
    "#         cgr_matrix.to_csv(sample_files[i], index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "3a2697c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_kenya = preprocessing(snp_list=\"../data/snp.lists\",\n",
    "                         df_bio=\"../accessions/kenya/kenya.tsv\",\n",
    "                         df_sra=\"../pipeline/run-2023-08-11-salmonellaml/kenya_biosamples.txt\",\n",
    "                         country=\"kenya\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "73935c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_tz = preprocessing(snp_list=\"../data/snp.lists\",\n",
    "                      df_bio=\"../work/5a/c62ab92992b24ec6c72c7c94c7903d/tanzania_biosamples.txt\",\n",
    "                      df_sra=\"../pipeline/run-2023-08-11-salmonellaml/sra_tanzania_biosamples.txt\",\n",
    "                      country=\"tanzania\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "f3ff2e81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>genomic_position</th>\n",
       "      <th>sra_accession</th>\n",
       "      <th>country</th>\n",
       "      <th>22</th>\n",
       "      <th>23</th>\n",
       "      <th>36</th>\n",
       "      <th>43</th>\n",
       "      <th>216</th>\n",
       "      <th>480</th>\n",
       "      <th>495</th>\n",
       "      <th>534</th>\n",
       "      <th>...</th>\n",
       "      <th>4841673</th>\n",
       "      <th>4841793</th>\n",
       "      <th>4844655</th>\n",
       "      <th>4851050</th>\n",
       "      <th>4854011</th>\n",
       "      <th>4854114</th>\n",
       "      <th>4854186</th>\n",
       "      <th>4854195</th>\n",
       "      <th>4854340</th>\n",
       "      <th>4855889</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR11005771</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR11005772</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR11005773</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR11005774</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR11005775</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>SRR8980095</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>SRR8980096</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>SRR8980097</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>SRR8980099</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>SRR8980101</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>N</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 331961 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "genomic_position sra_accession   country 22 23 36 43 216 480 495 534  ...  \\\n",
       "0                  SRR11005771  tanzania  G  C  G  C   T   G   C   C  ...   \n",
       "1                  SRR11005772  tanzania  G  C  G  C   T   G   C   C  ...   \n",
       "2                  SRR11005773  tanzania  G  C  G  C   N   G   C   C  ...   \n",
       "3                  SRR11005774  tanzania  N  N  N  N   N   G   N   C  ...   \n",
       "4                  SRR11005775  tanzania  N  N  N  N   N   G   N   C  ...   \n",
       "..                         ...       ... .. .. .. ..  ..  ..  ..  ..  ...   \n",
       "71                  SRR8980095  tanzania  N  N  N  N   N   G   N   C  ...   \n",
       "72                  SRR8980096  tanzania  N  N  N  N   N   G   N   C  ...   \n",
       "73                  SRR8980097  tanzania  G  C  N  N   N   G   N   C  ...   \n",
       "74                  SRR8980099  tanzania  N  N  N  N   N   G   N   C  ...   \n",
       "75                  SRR8980101  tanzania  N  C  N  N   N   G   N   C  ...   \n",
       "\n",
       "genomic_position 4841673 4841793 4844655 4851050 4854011 4854114 4854186  \\\n",
       "0                      N       N       N       N       N       N       N   \n",
       "1                      N       N       N       N       N       N       N   \n",
       "2                      N       N       N       N       N       N       N   \n",
       "3                      N       N       N       N       N       N       N   \n",
       "4                      N       N       N       N       N       N       N   \n",
       "..                   ...     ...     ...     ...     ...     ...     ...   \n",
       "71                     N       N       N       N       N       N       N   \n",
       "72                     N       N       N       N       N       N       N   \n",
       "73                     N       N       N       N       N       N       N   \n",
       "74                     N       N       N       N       N       N       N   \n",
       "75                     A       A       C       C       G       G       C   \n",
       "\n",
       "genomic_position 4854195 4854340 4855889  \n",
       "0                      N       N       N  \n",
       "1                      N       N       N  \n",
       "2                      N       N       N  \n",
       "3                      N       N       N  \n",
       "4                      N       N       N  \n",
       "..                   ...     ...     ...  \n",
       "71                     N       N       N  \n",
       "72                     N       N       N  \n",
       "73                     N       N       N  \n",
       "74                     N       N       N  \n",
       "75                     A       T       C  \n",
       "\n",
       "[76 rows x 331961 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "7678c18f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# inspect the constant columns\n",
    "# Step 2: Remove constant columns\n",
    "def non_constant(df):\n",
    "    constant_cols = df.columns[df.nunique() ==1]\n",
    "#     constant = df[constant_cols]\n",
    "    constant\n",
    "    return constant"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d69342b6",
   "metadata": {},
   "source": [
    "# Non constant Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "e2e27392",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>genomic_position</th>\n",
       "      <th>136569</th>\n",
       "      <th>185378</th>\n",
       "      <th>250051</th>\n",
       "      <th>277015</th>\n",
       "      <th>286832</th>\n",
       "      <th>410830</th>\n",
       "      <th>441536</th>\n",
       "      <th>459590</th>\n",
       "      <th>549075</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>132 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "genomic_position 136569 185378 250051 277015 286832 410830 441536 459590  \\\n",
       "0                     G      G      C      C      A      C      T      C   \n",
       "1                     G      G      C      C      A      C      T      C   \n",
       "2                     G      G      C      C      A      C      T      C   \n",
       "3                     G      G      C      C      A      C      T      C   \n",
       "4                     G      G      C      C      A      C      T      C   \n",
       "..                  ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "127                   G      G      C      C      A      C      T      C   \n",
       "128                   G      G      C      C      A      C      T      C   \n",
       "129                   G      G      C      C      A      C      T      C   \n",
       "130                   G      G      C      C      A      C      T      C   \n",
       "131                   G      G      C      C      A      C      T      C   \n",
       "\n",
       "genomic_position 549075  \n",
       "0                     C  \n",
       "1                     C  \n",
       "2                     C  \n",
       "3                     C  \n",
       "4                     C  \n",
       "..                  ...  \n",
       "127                   C  \n",
       "128                   C  \n",
       "129                   C  \n",
       "130                   C  \n",
       "131                   C  \n",
       "\n",
       "[132 rows x 9 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_constant(df_tz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "9c9cecae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>genomic_position</th>\n",
       "      <th>country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>kenya</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "genomic_position country\n",
       "0                  kenya\n",
       "1                  kenya\n",
       "2                  kenya\n",
       "3                  kenya\n",
       "4                  kenya\n",
       "..                   ...\n",
       "71                 kenya\n",
       "72                 kenya\n",
       "73                 kenya\n",
       "74                 kenya\n",
       "75                 kenya\n",
       "\n",
       "[76 rows x 1 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "non_constant(df_kenya)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0ea7001",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "56394613",
   "metadata": {},
   "source": [
    "# Nigeria\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "decf8478",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Organism group</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Isolate identifiers</th>\n",
       "      <th>Serovar</th>\n",
       "      <th>Isolate</th>\n",
       "      <th>Create date</th>\n",
       "      <th>Location</th>\n",
       "      <th>Isolation source</th>\n",
       "      <th>Isolation type</th>\n",
       "      <th>SNP cluster</th>\n",
       "      <th>Min-same</th>\n",
       "      <th>Min-diff</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>AMR genotypes</th>\n",
       "      <th>Computed types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>RKS304_601-84</td>\n",
       "      <td>BCW_2606,\"CFSAN005617\",\"RKS304_601-84\",\"SRS536...</td>\n",
       "      <td>1,40:z35:-</td>\n",
       "      <td>PDT000002955.5</td>\n",
       "      <td>2020-04-27T12:13:59Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>food</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN02368165</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=40:z35:-,serotype=[1],40:z35:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>33(3)</td>\n",
       "      <td>33(3),\"BCW_2809\",\"SRS538850\"</td>\n",
       "      <td>Uganda</td>\n",
       "      <td>PDT000003053.4</td>\n",
       "      <td>2015-12-12T15:21:45Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Human, faeces</td>\n",
       "      <td>clinical</td>\n",
       "      <td>PDS000001850.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SAMN02368677</td>\n",
       "      <td>GCA_010639825.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=3,10:l,z13:1,5,serotype=Uganda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>FMA0095</td>\n",
       "      <td>FMA0095,\"SRS565883\"</td>\n",
       "      <td>Ealing</td>\n",
       "      <td>PDT000003898.3</td>\n",
       "      <td>2015-02-11T02:07:39Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>melon seed</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN02345045</td>\n",
       "      <td>GCA_009997545.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=35:g,m,s:-,serotype=Ealing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>BCW_2640</td>\n",
       "      <td>9(3),\"BCW_2640\",\"SRS533427\"</td>\n",
       "      <td>Ank</td>\n",
       "      <td>PDT000047946.4</td>\n",
       "      <td>2015-07-23T05:10:39Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Cattle liver</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN02368530</td>\n",
       "      <td>GCA_010369405.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=28:k:e,n,z15,serotype=Ank</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>FDA904406 1-1</td>\n",
       "      <td>CFSAN032276,\"FDA00008919\",\"FDA904406 1-1\",\"SRS...</td>\n",
       "      <td>Pomona</td>\n",
       "      <td>PDT000060513.2</td>\n",
       "      <td>2015-04-18T14:10:49Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>ground chili pepper</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>PDS000031411.1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN03483118</td>\n",
       "      <td>GCA_011409935.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=28:y:1,7,serotype=Pomona</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       #Organism group         Strain  \\\n",
       "0  Salmonella enterica  RKS304_601-84   \n",
       "1  Salmonella enterica          33(3)   \n",
       "2  Salmonella enterica        FMA0095   \n",
       "3  Salmonella enterica       BCW_2640   \n",
       "4  Salmonella enterica  FDA904406 1-1   \n",
       "\n",
       "                                 Isolate identifiers     Serovar  \\\n",
       "0  BCW_2606,\"CFSAN005617\",\"RKS304_601-84\",\"SRS536...  1,40:z35:-   \n",
       "1                       33(3),\"BCW_2809\",\"SRS538850\"      Uganda   \n",
       "2                                FMA0095,\"SRS565883\"      Ealing   \n",
       "3                        9(3),\"BCW_2640\",\"SRS533427\"         Ank   \n",
       "4  CFSAN032276,\"FDA00008919\",\"FDA904406 1-1\",\"SRS...      Pomona   \n",
       "\n",
       "          Isolate           Create date Location     Isolation source  \\\n",
       "0  PDT000002955.5  2020-04-27T12:13:59Z  Nigeria                 food   \n",
       "1  PDT000003053.4  2015-12-12T15:21:45Z  Nigeria        Human, faeces   \n",
       "2  PDT000003898.3  2015-02-11T02:07:39Z  Nigeria           melon seed   \n",
       "3  PDT000047946.4  2015-07-23T05:10:39Z  Nigeria         Cattle liver   \n",
       "4  PDT000060513.2  2015-04-18T14:10:49Z  Nigeria  ground chili pepper   \n",
       "\n",
       "        Isolation type     SNP cluster  Min-same  Min-diff     BioSample  \\\n",
       "0  environmental/other             NaN       NaN       NaN  SAMN02368165   \n",
       "1             clinical  PDS000001850.5       NaN       0.0  SAMN02368677   \n",
       "2  environmental/other             NaN       NaN       NaN  SAMN02345045   \n",
       "3  environmental/other             NaN       NaN       NaN  SAMN02368530   \n",
       "4  environmental/other  PDS000031411.1       1.0       NaN  SAMN03483118   \n",
       "\n",
       "          Assembly                AMR genotypes  \\\n",
       "0              NaN                mdsB=COMPLETE   \n",
       "1  GCA_010639825.1  mdsA=COMPLETE,mdsB=COMPLETE   \n",
       "2  GCA_009997545.1  mdsA=COMPLETE,mdsB=COMPLETE   \n",
       "3  GCA_010369405.1  mdsA=COMPLETE,mdsB=COMPLETE   \n",
       "4  GCA_011409935.1  mdsA=COMPLETE,mdsB=COMPLETE   \n",
       "\n",
       "                                   Computed types  \n",
       "0  antigen_formula=40:z35:-,serotype=[1],40:z35:-  \n",
       "1  antigen_formula=3,10:l,z13:1,5,serotype=Uganda  \n",
       "2      antigen_formula=35:g,m,s:-,serotype=Ealing  \n",
       "3       antigen_formula=28:k:e,n,z15,serotype=Ank  \n",
       "4        antigen_formula=28:y:1,7,serotype=Pomona  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Nigeria\n",
    "# load data\n",
    "df_nigeria=pd.read_table(\"../data/nigeria.tsv\")\n",
    "df_nigeria.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "e23d9081",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Organism group</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Isolate identifiers</th>\n",
       "      <th>Serovar</th>\n",
       "      <th>Isolate</th>\n",
       "      <th>Create date</th>\n",
       "      <th>Location</th>\n",
       "      <th>Isolation source</th>\n",
       "      <th>Isolation type</th>\n",
       "      <th>SNP cluster</th>\n",
       "      <th>Min-same</th>\n",
       "      <th>Min-diff</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>AMR genotypes</th>\n",
       "      <th>Computed types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>BCW_2822</td>\n",
       "      <td>66(3),\"BCW_2822\",\"SRS528782\"</td>\n",
       "      <td>Wilhelmsburg</td>\n",
       "      <td>PDT000197317.1</td>\n",
       "      <td>2017-03-22T14:56:21Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Poultry small intenstine</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN02368689</td>\n",
       "      <td>GCA_002062955.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=4:z38:-,serotype=Wilhelmsburg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>BCW_2817</td>\n",
       "      <td>BCW_2817,\"KR 16\",\"SRS865334\"</td>\n",
       "      <td>Virchow</td>\n",
       "      <td>PDT000197326.1</td>\n",
       "      <td>2017-03-22T15:02:25Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>chicken</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>PDS000031736.79</td>\n",
       "      <td>22.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>SAMN02368684</td>\n",
       "      <td>GCA_002063155.1</td>\n",
       "      <td>dfrA5=COMPLETE,gyrA_S83Y=POINT,mdsA=COMPLETE,m...</td>\n",
       "      <td>antigen_formula=7:r:1,2,serotype=Virchow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>BCW_2794</td>\n",
       "      <td>BCW_2794,\"KR 225\",\"SRS528772\"</td>\n",
       "      <td>Bredeney</td>\n",
       "      <td>PDT000197354.1</td>\n",
       "      <td>2017-03-22T15:10:09Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>chicken</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN02368666</td>\n",
       "      <td>GCA_002063755.1</td>\n",
       "      <td>blaTEM-1=COMPLETE,gyrA_D87N=POINT,mdsA=COMPLET...</td>\n",
       "      <td>antigen_formula=4:l,v:1,7,serotype=Bredeney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>BCW_2750</td>\n",
       "      <td>BCW_2750,\"KR167\",\"SRS528743\"</td>\n",
       "      <td>Onireke</td>\n",
       "      <td>PDT000197377.1</td>\n",
       "      <td>2017-03-22T15:18:10Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>chicken</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>PDS000001391.75</td>\n",
       "      <td>8.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>SAMN02368627</td>\n",
       "      <td>GCA_002064225.1</td>\n",
       "      <td>aadA1=COMPLETE,dfrA1=COMPLETE,gyrA_D87G=POINT,...</td>\n",
       "      <td>antigen_formula=4:b:1,2,serotype=Paratyphi B v...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>BCW_2729</td>\n",
       "      <td>BCW_2729,\"KR 84\",\"SRS528729\"</td>\n",
       "      <td>Mbandaka</td>\n",
       "      <td>PDT000197390.1</td>\n",
       "      <td>2017-03-22T15:21:50Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>chicken</td>\n",
       "      <td>environmental/other</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMN02368606</td>\n",
       "      <td>GCA_002064495.1</td>\n",
       "      <td>aadA1=COMPLETE,dfrA15=COMPLETE,gyrA_D87Y=POINT...</td>\n",
       "      <td>antigen_formula=7:z10:e,n,z15,serotype=Mbandaka</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D38,\"ERS4597431\",\"Salmonella\"</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT001291149.1</td>\n",
       "      <td>2022-04-20T13:38:33Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Poultry vendoors and farms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDS000019581.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA6869863</td>\n",
       "      <td>GCA_023070735.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE,qnrB19=COMPLETE</td>\n",
       "      <td>antigen_formula=4:z10:1,5,serotype=Ituri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D39,\"ERS4597432\",\"Salmonella\"</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT001291150.1</td>\n",
       "      <td>2022-04-20T13:38:33Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Poultry vendoors and farms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDS000108531.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA6869864</td>\n",
       "      <td>GCA_023070715.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=7:r:1,2,serotype=Virchow</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D40,\"ERS4597433\",\"Salmonella\"</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT001291151.1</td>\n",
       "      <td>2022-04-20T13:38:33Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Poultry vendoors and farms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDS000108534.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA6869865</td>\n",
       "      <td>GCA_023070695.1</td>\n",
       "      <td>fosA7=COMPLETE,fosB=PARTIAL_END_OF_CONTIG,mdsA...</td>\n",
       "      <td>antigen_formula=4:z4,z23:1,2,serotype=Stanleyv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D41,\"ERS4597434\",\"Salmonella\"</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT001291152.1</td>\n",
       "      <td>2022-04-20T13:38:34Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Poultry vendoors and farms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA6869866</td>\n",
       "      <td>GCA_023070635.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=7:g,s,t:-,serotype=Menston</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>686</th>\n",
       "      <td>Salmonella enterica</td>\n",
       "      <td>NaN</td>\n",
       "      <td>D42,\"ERS4597435\",\"Salmonella\"</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDT001291153.1</td>\n",
       "      <td>2022-04-20T13:38:34Z</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Poultry vendoors and farms</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAMEA6869867</td>\n",
       "      <td>GCA_023070675.1</td>\n",
       "      <td>mdsA=COMPLETE,mdsB=COMPLETE</td>\n",
       "      <td>antigen_formula=-:m,t:-,serotype=I -:m,t:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #Organism group    Strain            Isolate identifiers  \\\n",
       "88   Salmonella enterica  BCW_2822   66(3),\"BCW_2822\",\"SRS528782\"   \n",
       "89   Salmonella enterica  BCW_2817   BCW_2817,\"KR 16\",\"SRS865334\"   \n",
       "92   Salmonella enterica  BCW_2794  BCW_2794,\"KR 225\",\"SRS528772\"   \n",
       "94   Salmonella enterica  BCW_2750   BCW_2750,\"KR167\",\"SRS528743\"   \n",
       "95   Salmonella enterica  BCW_2729   BCW_2729,\"KR 84\",\"SRS528729\"   \n",
       "..                   ...       ...                            ...   \n",
       "682  Salmonella enterica       NaN  D38,\"ERS4597431\",\"Salmonella\"   \n",
       "683  Salmonella enterica       NaN  D39,\"ERS4597432\",\"Salmonella\"   \n",
       "684  Salmonella enterica       NaN  D40,\"ERS4597433\",\"Salmonella\"   \n",
       "685  Salmonella enterica       NaN  D41,\"ERS4597434\",\"Salmonella\"   \n",
       "686  Salmonella enterica       NaN  D42,\"ERS4597435\",\"Salmonella\"   \n",
       "\n",
       "          Serovar         Isolate           Create date Location  \\\n",
       "88   Wilhelmsburg  PDT000197317.1  2017-03-22T14:56:21Z  Nigeria   \n",
       "89        Virchow  PDT000197326.1  2017-03-22T15:02:25Z  Nigeria   \n",
       "92       Bredeney  PDT000197354.1  2017-03-22T15:10:09Z  Nigeria   \n",
       "94        Onireke  PDT000197377.1  2017-03-22T15:18:10Z  Nigeria   \n",
       "95       Mbandaka  PDT000197390.1  2017-03-22T15:21:50Z  Nigeria   \n",
       "..            ...             ...                   ...      ...   \n",
       "682           NaN  PDT001291149.1  2022-04-20T13:38:33Z  Nigeria   \n",
       "683           NaN  PDT001291150.1  2022-04-20T13:38:33Z  Nigeria   \n",
       "684           NaN  PDT001291151.1  2022-04-20T13:38:33Z  Nigeria   \n",
       "685           NaN  PDT001291152.1  2022-04-20T13:38:34Z  Nigeria   \n",
       "686           NaN  PDT001291153.1  2022-04-20T13:38:34Z  Nigeria   \n",
       "\n",
       "               Isolation source       Isolation type      SNP cluster  \\\n",
       "88     Poultry small intenstine  environmental/other              NaN   \n",
       "89                      chicken  environmental/other  PDS000031736.79   \n",
       "92                      chicken  environmental/other              NaN   \n",
       "94                      chicken  environmental/other  PDS000001391.75   \n",
       "95                      chicken  environmental/other              NaN   \n",
       "..                          ...                  ...              ...   \n",
       "682  Poultry vendoors and farms                  NaN   PDS000019581.6   \n",
       "683  Poultry vendoors and farms                  NaN   PDS000108531.2   \n",
       "684  Poultry vendoors and farms                  NaN   PDS000108534.1   \n",
       "685  Poultry vendoors and farms                  NaN              NaN   \n",
       "686  Poultry vendoors and farms                  NaN              NaN   \n",
       "\n",
       "     Min-same  Min-diff     BioSample         Assembly  \\\n",
       "88        NaN       NaN  SAMN02368689  GCA_002062955.1   \n",
       "89       22.0      16.0  SAMN02368684  GCA_002063155.1   \n",
       "92        NaN       NaN  SAMN02368666  GCA_002063755.1   \n",
       "94        8.0      14.0  SAMN02368627  GCA_002064225.1   \n",
       "95        NaN       NaN  SAMN02368606  GCA_002064495.1   \n",
       "..        ...       ...           ...              ...   \n",
       "682       NaN       NaN  SAMEA6869863  GCA_023070735.1   \n",
       "683       NaN       NaN  SAMEA6869864  GCA_023070715.1   \n",
       "684       NaN       NaN  SAMEA6869865  GCA_023070695.1   \n",
       "685       NaN       NaN  SAMEA6869866  GCA_023070635.1   \n",
       "686       NaN       NaN  SAMEA6869867  GCA_023070675.1   \n",
       "\n",
       "                                         AMR genotypes  \\\n",
       "88                         mdsA=COMPLETE,mdsB=COMPLETE   \n",
       "89   dfrA5=COMPLETE,gyrA_S83Y=POINT,mdsA=COMPLETE,m...   \n",
       "92   blaTEM-1=COMPLETE,gyrA_D87N=POINT,mdsA=COMPLET...   \n",
       "94   aadA1=COMPLETE,dfrA1=COMPLETE,gyrA_D87G=POINT,...   \n",
       "95   aadA1=COMPLETE,dfrA15=COMPLETE,gyrA_D87Y=POINT...   \n",
       "..                                                 ...   \n",
       "682        mdsA=COMPLETE,mdsB=COMPLETE,qnrB19=COMPLETE   \n",
       "683                        mdsA=COMPLETE,mdsB=COMPLETE   \n",
       "684  fosA7=COMPLETE,fosB=PARTIAL_END_OF_CONTIG,mdsA...   \n",
       "685                        mdsA=COMPLETE,mdsB=COMPLETE   \n",
       "686                        mdsA=COMPLETE,mdsB=COMPLETE   \n",
       "\n",
       "                                        Computed types  \n",
       "88       antigen_formula=4:z38:-,serotype=Wilhelmsburg  \n",
       "89            antigen_formula=7:r:1,2,serotype=Virchow  \n",
       "92         antigen_formula=4:l,v:1,7,serotype=Bredeney  \n",
       "94   antigen_formula=4:b:1,2,serotype=Paratyphi B v...  \n",
       "95     antigen_formula=7:z10:e,n,z15,serotype=Mbandaka  \n",
       "..                                                 ...  \n",
       "682           antigen_formula=4:z10:1,5,serotype=Ituri  \n",
       "683           antigen_formula=7:r:1,2,serotype=Virchow  \n",
       "684  antigen_formula=4:z4,z23:1,2,serotype=Stanleyv...  \n",
       "685         antigen_formula=7:g,s,t:-,serotype=Menston  \n",
       "686         antigen_formula=-:m,t:-,serotype=I -:m,t:-  \n",
       "\n",
       "[84 rows x 16 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load the data\n",
    "# Filter the DataFrame\n",
    "filtered_df = df_nigeria[(df_nigeria['Isolation source'].notnull()) & (~df_nigeria['Isolation source'].isin(['feces', 'stool']))]\n",
    " # Assuming df_nigeria is your original DataFrame\n",
    "filtered_poultry = df_nigeria[df_nigeria['Isolation source'].isin(['poultry feed','Poultry vendoors and farms','chicken heart','Poultry small intenstine','Poultry carcass','chicken liver','Chicken Spleen','chicken','poultry water'])]                             \n",
    "# Display the filtered DataFrame\n",
    "filtered_poultry.to_csv(\"../accessions/nigeria/filtered_nigeria_nigeria_poultry_data.tsv\", sep='\\t', index=False)\n",
    "filtered_poultry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "7b545d65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['food',\n",
       " 'Human, faeces',\n",
       " 'melon seed',\n",
       " 'Cattle liver',\n",
       " 'ground chili pepper',\n",
       " 'ground melon seed egusi',\n",
       " 'ground melon seed egusi',\n",
       " 'ground chili pepper',\n",
       " 'melon seed',\n",
       " 'melon seed',\n",
       " 'melon seed',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'rectal swab (Bos taurus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'cloacal swab (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'tissue (Gallus gallus)',\n",
       " 'dried ginger split',\n",
       " 'frozen snail',\n",
       " 'frozen snail',\n",
       " 'vegetable broth',\n",
       " 'vegetable broth',\n",
       " 'dried crayfish',\n",
       " 'dried crayfish',\n",
       " 'melon',\n",
       " 'Poultry small intenstine',\n",
       " 'chicken',\n",
       " 'vegetable',\n",
       " 'Poultry spleen',\n",
       " 'chicken',\n",
       " 'Human Blood',\n",
       " 'chicken',\n",
       " 'chicken',\n",
       " 'Lizard Faeces',\n",
       " 'water',\n",
       " 'Human Blood',\n",
       " 'Human blood',\n",
       " 'chicken',\n",
       " 'Human Blood',\n",
       " 'chicken',\n",
       " 'Lizard, faeces',\n",
       " 'chicken',\n",
       " 'Spinach',\n",
       " 'chicken',\n",
       " 'fish intestine',\n",
       " 'Human Blood',\n",
       " 'Lizard faeces',\n",
       " 'Human blood',\n",
       " 'Rodent faeces',\n",
       " 'Camel spleen',\n",
       " 'chicken',\n",
       " 'Cattle liver',\n",
       " 'Camel intentine',\n",
       " 'ground red pepper',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'pig stool',\n",
       " 'chicken stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'chicken stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'cattle stool (Bos taurus)',\n",
       " 'stool (Bos taurus)',\n",
       " 'pig stool',\n",
       " 'pig stool',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'cloacal swab',\n",
       " 'feces pig',\n",
       " 'poultry feed',\n",
       " 'intestine cattle',\n",
       " 'poultry litter',\n",
       " 'Chicken Spleen',\n",
       " 'poultry feed',\n",
       " 'intestine cattle',\n",
       " 'milk',\n",
       " 'Chicken Spleen',\n",
       " 'cloacal swab',\n",
       " 'feces pig',\n",
       " 'milk',\n",
       " 'cloacal swab',\n",
       " 'milk',\n",
       " 'litter swab',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'stool (Bos taurus)',\n",
       " 'blood',\n",
       " 'blood',\n",
       " 'blood',\n",
       " 'milk',\n",
       " 'feces pig',\n",
       " 'Chicken Spleen',\n",
       " 'feces pig',\n",
       " 'chicken liver',\n",
       " 'feces pig',\n",
       " 'Chicken Spleen',\n",
       " 'chicken liver',\n",
       " 'cattle',\n",
       " 'chicken caecum',\n",
       " 'feces pig',\n",
       " 'cattle lung',\n",
       " 'cloacal swab',\n",
       " 'cloacal swab',\n",
       " 'cloacal swab',\n",
       " 'chicken heart',\n",
       " 'cloacal swab',\n",
       " 'poultry feed',\n",
       " 'poultry feed',\n",
       " 'chicken liver',\n",
       " 'cloacal swab',\n",
       " 'poultry water',\n",
       " 'poultry feed',\n",
       " 'Chicken Spleen',\n",
       " 'cloacal swab',\n",
       " 'cloacal swab',\n",
       " 'cloacal swab',\n",
       " 'cloacal swab',\n",
       " 'cloacal swab',\n",
       " 'cloacal swab',\n",
       " 'cloacal swab',\n",
       " 'chicken liver',\n",
       " 'cloacal swab',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'cloacal swab',\n",
       " 'chicken liver',\n",
       " 'cattle lung',\n",
       " 'intestine cattle',\n",
       " 'feces pig',\n",
       " 'catte',\n",
       " 'feces pig',\n",
       " 'pig stool',\n",
       " 'poultry feed',\n",
       " 'litter swab',\n",
       " 'chicken liver',\n",
       " 'chicken heart',\n",
       " 'chicken heart',\n",
       " 'feces pig',\n",
       " 'cattle',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'poultry feed',\n",
       " 'cattle liver',\n",
       " 'feces pig',\n",
       " 'chicken liver',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'feces pig',\n",
       " 'clinical isolate',\n",
       " 'clinical isolate',\n",
       " 'clinical isolate',\n",
       " 'clinical isolate',\n",
       " 'clinical isolate',\n",
       " 'clinical isolate',\n",
       " 'melon powder',\n",
       " 'melon powder',\n",
       " 'clinical isolate',\n",
       " 'clinical isolate',\n",
       " 'clinical isolate',\n",
       " 'clinical isolate',\n",
       " 'pepper',\n",
       " 'pepper',\n",
       " 'egusi melon seed',\n",
       " 'egusi melon seed',\n",
       " 'Poultry',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Canis familiaris)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'bitter leaves',\n",
       " 'bitter leaves',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'feces (Agama agama)',\n",
       " 'ground melon seeds',\n",
       " 'ground melon seeds',\n",
       " 'ground melon seeds',\n",
       " 'ground melon',\n",
       " 'ground melon',\n",
       " 'human',\n",
       " 'human',\n",
       " 'cattle',\n",
       " 'Environment',\n",
       " 'Environment',\n",
       " 'Poultry feed',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Human sludge',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry carcass',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'Poultry vendoors and farms',\n",
       " 'blood',\n",
       " 'urine',\n",
       " 'blood',\n",
       " 'Pus Osteitis',\n",
       " 'blood',\n",
       " 'Snail',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'Beef liver',\n",
       " 'beef',\n",
       " 'Beef liver',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'liver',\n",
       " 'beef',\n",
       " 'liver',\n",
       " 'liver',\n",
       " 'liver',\n",
       " 'liver',\n",
       " 'beef',\n",
       " 'GOAT MEAT',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'liver',\n",
       " 'liver',\n",
       " 'liver',\n",
       " 'liver',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'grounded melon powder',\n",
       " 'grounded melon powder',\n",
       " 'grounded melon powder',\n",
       " 'blood',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'liver',\n",
       " 'liver',\n",
       " 'liver',\n",
       " 'beef',\n",
       " 'liver',\n",
       " 'Dog feces',\n",
       " 'beef',\n",
       " 'liver',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'beef',\n",
       " 'liver',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated',\n",
       " 'Clinical/host-associated']"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# One hot encoding\n",
    "list(filtered_df[\"Isolation source\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "806e7c86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replicate for trial \n",
    "\n",
    "\n",
    "# Create a new dataset with 500 rows\n",
    "# Replicate the data 5 times (total 500 rows)\n",
    "for _ in range(6):\n",
    "    replicated_data = pd.concat([df, df], ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d84df478",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Biosample</th>\n",
       "      <th>sra_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN20456381</td>\n",
       "      <td>SRR15403520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN20456395</td>\n",
       "      <td>SRR15369306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN20456385</td>\n",
       "      <td>SRR15369305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN20456390</td>\n",
       "      <td>SRR15369303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN20456392</td>\n",
       "      <td>SRR15369302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>SAMN03576959</td>\n",
       "      <td>SRR2163255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>SAMN03576956</td>\n",
       "      <td>SRR2163254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>SAMN03576963</td>\n",
       "      <td>SRR2163247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>SAMN03576962</td>\n",
       "      <td>SRR2163246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>SAMN03576958</td>\n",
       "      <td>SRR2163243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>268 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Biosample sra_accession\n",
       "0    SAMN20456381   SRR15403520\n",
       "1    SAMN20456395   SRR15369306\n",
       "2    SAMN20456385   SRR15369305\n",
       "3    SAMN20456390   SRR15369303\n",
       "4    SAMN20456392   SRR15369302\n",
       "..            ...           ...\n",
       "263  SAMN03576959    SRR2163255\n",
       "264  SAMN03576956    SRR2163254\n",
       "265  SAMN03576963    SRR2163247\n",
       "266  SAMN03576962    SRR2163246\n",
       "267  SAMN03576958    SRR2163243\n",
       "\n",
       "[268 rows x 2 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "replicated_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2b591360",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Biosample</th>\n",
       "      <th>sra_accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMN20456381</td>\n",
       "      <td>SRR15403520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMN20456395</td>\n",
       "      <td>SRR15369306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMN20456385</td>\n",
       "      <td>SRR15369305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMN20456390</td>\n",
       "      <td>SRR15369303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN20456392</td>\n",
       "      <td>SRR15369302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>SAMN03576959</td>\n",
       "      <td>SRR2163255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>264</th>\n",
       "      <td>SAMN03576956</td>\n",
       "      <td>SRR2163254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>SAMN03576963</td>\n",
       "      <td>SRR2163247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>266</th>\n",
       "      <td>SAMN03576962</td>\n",
       "      <td>SRR2163246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267</th>\n",
       "      <td>SAMN03576958</td>\n",
       "      <td>SRR2163243</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>268 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Biosample sra_accession\n",
       "0    SAMN20456381   SRR15403520\n",
       "1    SAMN20456395   SRR15369306\n",
       "2    SAMN20456385   SRR15369305\n",
       "3    SAMN20456390   SRR15369303\n",
       "4    SAMN20456392   SRR15369302\n",
       "..            ...           ...\n",
       "263  SAMN03576959    SRR2163255\n",
       "264  SAMN03576956    SRR2163254\n",
       "265  SAMN03576963    SRR2163247\n",
       "266  SAMN03576962    SRR2163246\n",
       "267  SAMN03576958    SRR2163243\n",
       "\n",
       "[268 rows x 2 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Step 2: Remove constant columns\n",
    "non_constant_cols = replicated_data.columns[replicated_data.nunique() > 1]\n",
    "replicated_data = replicated_data[non_constant_cols]\n",
    "replicated_data.reindex()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3f39f2ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Biosample_SAMN02843543</th>\n",
       "      <th>Biosample_SAMN02844775</th>\n",
       "      <th>Biosample_SAMN03576951</th>\n",
       "      <th>Biosample_SAMN03576952</th>\n",
       "      <th>Biosample_SAMN03576953</th>\n",
       "      <th>Biosample_SAMN03576954</th>\n",
       "      <th>Biosample_SAMN03576955</th>\n",
       "      <th>Biosample_SAMN03576956</th>\n",
       "      <th>Biosample_SAMN03576958</th>\n",
       "      <th>Biosample_SAMN03576959</th>\n",
       "      <th>...</th>\n",
       "      <th>sra_accession_SRR3747672</th>\n",
       "      <th>sra_accession_SRR3924533</th>\n",
       "      <th>sra_accession_SRR3924536</th>\n",
       "      <th>sra_accession_SRR4123985</th>\n",
       "      <th>sra_accession_SRR4123986</th>\n",
       "      <th>sra_accession_SRR4243040</th>\n",
       "      <th>sra_accession_SRR4733853</th>\n",
       "      <th>sra_accession_SRR4733857</th>\n",
       "      <th>sra_accession_SRR5972858</th>\n",
       "      <th>sra_accession_SRR6958628</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>134 rows × 266 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Biosample_SAMN02843543  Biosample_SAMN02844775  Biosample_SAMN03576951  \\\n",
       "0                     False                   False                   False   \n",
       "1                     False                   False                   False   \n",
       "2                     False                   False                   False   \n",
       "3                     False                   False                   False   \n",
       "4                     False                   False                   False   \n",
       "..                      ...                     ...                     ...   \n",
       "129                   False                   False                   False   \n",
       "130                   False                   False                   False   \n",
       "131                   False                   False                   False   \n",
       "132                   False                   False                   False   \n",
       "133                   False                   False                   False   \n",
       "\n",
       "     Biosample_SAMN03576952  Biosample_SAMN03576953  Biosample_SAMN03576954  \\\n",
       "0                     False                   False                   False   \n",
       "1                     False                   False                   False   \n",
       "2                     False                   False                   False   \n",
       "3                     False                   False                   False   \n",
       "4                     False                   False                   False   \n",
       "..                      ...                     ...                     ...   \n",
       "129                   False                   False                   False   \n",
       "130                   False                   False                   False   \n",
       "131                   False                   False                   False   \n",
       "132                   False                   False                   False   \n",
       "133                   False                   False                   False   \n",
       "\n",
       "     Biosample_SAMN03576955  Biosample_SAMN03576956  Biosample_SAMN03576958  \\\n",
       "0                     False                   False                   False   \n",
       "1                     False                   False                   False   \n",
       "2                     False                   False                   False   \n",
       "3                     False                   False                   False   \n",
       "4                     False                   False                   False   \n",
       "..                      ...                     ...                     ...   \n",
       "129                   False                   False                   False   \n",
       "130                   False                    True                   False   \n",
       "131                   False                   False                   False   \n",
       "132                   False                   False                   False   \n",
       "133                   False                   False                    True   \n",
       "\n",
       "     Biosample_SAMN03576959  ...  sra_accession_SRR3747672  \\\n",
       "0                     False  ...                     False   \n",
       "1                     False  ...                     False   \n",
       "2                     False  ...                     False   \n",
       "3                     False  ...                     False   \n",
       "4                     False  ...                     False   \n",
       "..                      ...  ...                       ...   \n",
       "129                    True  ...                     False   \n",
       "130                   False  ...                     False   \n",
       "131                   False  ...                     False   \n",
       "132                   False  ...                     False   \n",
       "133                   False  ...                     False   \n",
       "\n",
       "     sra_accession_SRR3924533  sra_accession_SRR3924536  \\\n",
       "0                       False                     False   \n",
       "1                       False                     False   \n",
       "2                       False                     False   \n",
       "3                       False                     False   \n",
       "4                       False                     False   \n",
       "..                        ...                       ...   \n",
       "129                     False                     False   \n",
       "130                     False                     False   \n",
       "131                     False                     False   \n",
       "132                     False                     False   \n",
       "133                     False                     False   \n",
       "\n",
       "     sra_accession_SRR4123985  sra_accession_SRR4123986  \\\n",
       "0                       False                     False   \n",
       "1                       False                     False   \n",
       "2                       False                     False   \n",
       "3                       False                     False   \n",
       "4                       False                     False   \n",
       "..                        ...                       ...   \n",
       "129                     False                     False   \n",
       "130                     False                     False   \n",
       "131                     False                     False   \n",
       "132                     False                     False   \n",
       "133                     False                     False   \n",
       "\n",
       "     sra_accession_SRR4243040  sra_accession_SRR4733853  \\\n",
       "0                       False                     False   \n",
       "1                       False                     False   \n",
       "2                       False                     False   \n",
       "3                       False                     False   \n",
       "4                       False                     False   \n",
       "..                        ...                       ...   \n",
       "129                     False                     False   \n",
       "130                     False                     False   \n",
       "131                     False                     False   \n",
       "132                     False                     False   \n",
       "133                     False                     False   \n",
       "\n",
       "     sra_accession_SRR4733857  sra_accession_SRR5972858  \\\n",
       "0                       False                     False   \n",
       "1                       False                     False   \n",
       "2                       False                     False   \n",
       "3                       False                     False   \n",
       "4                       False                     False   \n",
       "..                        ...                       ...   \n",
       "129                     False                     False   \n",
       "130                     False                     False   \n",
       "131                     False                     False   \n",
       "132                     False                     False   \n",
       "133                     False                     False   \n",
       "\n",
       "     sra_accession_SRR6958628  \n",
       "0                       False  \n",
       "1                       False  \n",
       "2                       False  \n",
       "3                       False  \n",
       "4                       False  \n",
       "..                        ...  \n",
       "129                     False  \n",
       "130                     False  \n",
       "131                     False  \n",
       "132                     False  \n",
       "133                     False  \n",
       "\n",
       "[134 rows x 266 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Step 2: Perform one-hot encoding for all categorical columns\n",
    "replicated_data_encoded = pd.get_dummies(df)\n",
    "replicated_data_encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "baa976ed",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'MinMaxScaler' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [117]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Step 3: Scale the data\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m scaler \u001b[38;5;241m=\u001b[39m \u001b[43mMinMaxScaler\u001b[49m()  \u001b[38;5;66;03m# or StandardScaler()\u001b[39;00m\n\u001b[1;32m      3\u001b[0m scaled_data \u001b[38;5;241m=\u001b[39m scaler\u001b[38;5;241m.\u001b[39mfit_transform(replicated_data)\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Convert the scaled numpy array back to a DataFrame with the original columns\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'MinMaxScaler' is not defined"
     ]
    }
   ],
   "source": [
    "# Step 3: Scale the data\n",
    "# from sklearn.preprocessing import MinMaxScaler\n",
    "scaler = MinMaxScaler()  # or StandardScaler()\n",
    "scaled_data = scaler.fit_transform(replicated_data)\n",
    "\n",
    "# Convert the scaled numpy array back to a DataFrame with the original columns\n",
    "scaled_data_df = pd.DataFrame(scaled_data, columns=non_constant_cols)\n",
    "scaled_data_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5ececcec",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import MinMaxScaler\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7e5d357",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
